### PR TITLE
docs: improve customization chapter

### PR DESCRIPTION
**Pull request by sennierer**

### CODE REVIEW

File: docs/source/customization.rst [hunk 1]
```diff
@@ -1,6 +1,9 @@
 Customization
 =============
 
+APIS is designed to be easily customizable. This section describes how you can
+customize the views and templates of the generic views that are shipped with
```
Comment (b1rger) [inline docs/source/customization.rst:5]:
```suggestion
customize the views and the templates of the views that are shipped with
```

File: docs/source/customization.rst [hunk 2]
```diff
@@ -1,6 +1,9 @@
...
+APIS by overwriting the classes and methods that are used by the generic views.
```
Comment (b1rger) [inline docs/source/customization.rst:6]:
```suggestion
APIS by injecting classes and methods that are then automatically used by the generic views.
```

File: docs/source/customization.rst [hunk 3]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+
+
+Extending autocomplete results
+------------------------------
+
+The autocomplete endpoint :py:class:`apis_core.generic.views.Autocomplete` can
+be extended to provide additional autocomplete results, that are not
+based on the default Django querysets. The autocomplete view always refers to a
+Django model. To extend the results for a specific model, you have to create an
+`ExternalAutocomplete` class, that is named after the model and resides in the
```
Comment (b1rger) [inline docs/source/customization.rst:160]:
```suggestion
``ExternalAutocomplete`` class, that is named after the model and resides in the
```

File: docs/source/customization.rst [hunk 4]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
 inherit from ``MyAbstractModel``, you can for example create an override table
 for all your models by creating a ``myproject.tables.MyAbstractModelTable``.
+
+
+Extending autocomplete results
+------------------------------
+
+The autocomplete endpoint :py:class:`apis_core.generic.views.Autocomplete` can
+be extended to provide additional autocomplete results, that are not
+based on the default Django querysets. The autocomplete view always refers to a
```
Comment (b1rger) [inline docs/source/customization.rst:158]: > The autocomplete view always refers to a Django model

It took me a while to figure out what that refers to. Given that *all* the views in the `generic.views` model use a model as a basis, I think this sentence can be removed (or moved to the top).

File: docs/source/customization.rst [hunk 5]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+Django model. To extend the results for a specific model, you have to create an
+`ExternalAutocomplete` class, that is named after the model and resides in the
+`querysets` module in the same app. So if you have an app called `myapp` with a
```
Comment (b1rger) [inline docs/source/customization.rst:161]:
```suggestion
``querysets`` module in the same app. So if you have an app called ``myapp`` with a
```

File: docs/source/customization.rst [hunk 6]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+`models.py`
```
Comment (b1rger) [inline docs/source/customization.rst:162]:
```suggestion
``models.py``
```

File: docs/source/customization.rst [hunk 7]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+
+.. code-block:: python
+
+   class Person(models.Model):
+        name = models.CharField(max_length=255)
+
+then the respective autocomplete class should reside in `myapp.querysets` and
```
Comment (b1rger) [inline docs/source/customization.rst:169]:
```suggestion
then the respective autocomplete class should reside in ``myapp.querysets`` and
```

File: docs/source/customization.rst [hunk 8]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+has to be called `PersonExternalAutocomplete`.
```
Comment (b1rger) [inline docs/source/customization.rst:170]:
```suggestion
has to be called ``PersonExternalAutocomplete``.
```

File: docs/source/customization.rst [hunk 9]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+            ... do something with the data
+            return data
+
+        def get_results(self, q):
+            with urllib.request.urlopen(f"https://some.uri.tld/search?q={q}") as f:
+                data = extract_results(json.loads(f.read()))
+                return results
+            return {}
+
+The class has to have a `get_results` method that receives a query as the first
```
Comment (b1rger) [inline docs/source/customization.rst:185]:
```suggestion
The class has to have a ``get_results`` method that receives a query as the first
```

File: docs/source/customization.rst [hunk 10]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+parameter and returns a result in the format, the `django-autocomplete-light`
```
Comment (b1rger) [inline docs/source/customization.rst:186]:
```suggestion
parameter and returns a result in the format, the `django-autocomplete-light <https://django-autocomplete-light.readthedocs.io/>`_
```

File: docs/source/customization.rst [hunk 11]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+module uses- this is a dict with the keys "id", "text" and "selected_text".
+
+
+Importing data from external resources
+--------------------------------------
+
+APIS provides the structure for easily importing data from external resources.
+One main component for this are `Importer` classes. They always belong to a
```
Comment (b1rger) [inline docs/source/customization.rst:194]:
```suggestion
One main component for this are ``Importer`` classes. They always belong to a
```

File: docs/source/customization.rst [hunk 12]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+Django Model, reside in the same app as the Djang model in the `importers`
```
Comment (b1rger) [inline docs/source/customization.rst:195]:
```suggestion
Django model, reside in the same app as the Django model in the ``importers``
```

File: docs/source/customization.rst [hunk 13]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+module and are named after the Django Model. So if you have an app called
```
Comment (b1rger) [inline docs/source/customization.rst:196]:
```suggestion
module and are named after the Django model. So if you have an app called
```

File: docs/source/customization.rst [hunk 14]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+`myapp` with a `models.py`
```
Comment (b1rger) [inline docs/source/customization.rst:197]:
```suggestion
``myapp`` with a ``models.py``
```

File: docs/source/customization.rst [hunk 15]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+
+.. code-block:: python
+
+   class Person(models.Model):
+        name = models.CharField(max_length=255)
+
+then the respective importer should reside in `myapp.importers` and has to be
```
Comment (b1rger) [inline docs/source/customization.rst:204]:
```suggestion
then the respective importer should reside in ``myapp.importers`` and has to be
```

File: docs/source/customization.rst [hunk 16]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+called `PersonImporter`.
```
Comment (b1rger) [inline docs/source/customization.rst:205]:
```suggestion
called ``PersonImporter``.
```

File: docs/source/customization.rst [hunk 17]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+
+An importer takes two arguments to instantiate: an `uri` and a `model`. The
```
Comment (b1rger) [inline docs/source/customization.rst:207]:
```suggestion
An importer takes two arguments to instantiate: an ``uri`` and a ``model``. The
```

File: docs/source/customization.rst [hunk 18]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+importers task is then to create a model instance from this URI, usually by
+fetching data from the URI, parsing it and extracting the needed fields.
+The instance should then be returned by the `create_instance` method of the
```
Comment (b1rger) [inline docs/source/customization.rst:210]:
```suggestion
The instance should then be returned by the ``create_instance`` method of the
```

File: docs/source/customization.rst [hunk 19]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+importer. There is :py:class:`apis_core.generic.importers.GenericImporter`
```
Comment (b1rger) [inline docs/source/customization.rst:211]:
```suggestion
importer. There is :py:class:`apis_core.generic.importers.GenericModelImporter`
```

File: docs/source/customization.rst [hunk 20]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+which you can inherit from.
```
Comment (b1rger) [inline docs/source/customization.rst:212]:
```suggestion
which you can inherit from. It is used by default of no other importer is defined for the model and it tries to do the right thing out of the box: it first looks if there is an RDF configuration for the URI and if that fails tries to parse the URI response as json.
```

File: docs/source/customization.rst [hunk 21]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+
+To use this logic in forms, there is
+:py:class:`apis_core.generic.forms.fields.ModelImportChoiceField` which is
+based on `django.forms.ModelChoiceField`. It checks if the passed value starts
```
Comment (b1rger) [inline docs/source/customization.rst:216]:
```suggestion
based on `django.forms.ModelChoiceField <https://docs.djangoproject.com/en/5.0/ref/forms/fields/#modelchoicefield>`_. It checks if the passed value starts
```

File: docs/source/customization.rst [hunk 22]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
...
+with `http` and if so, it uses the importer that fits the model and uses it to
```
Comment (b1rger) [inline docs/source/customization.rst:217]:
```suggestion
with ``http`` and if so, it uses the importer that fits the model and uses it to
```

**Review by b1rger — CHANGES_REQUESTED** [review]

File: docs/source/customization.rst [hunk 23]
```diff
@@ -145,3 +148,71 @@ look for possible overrides using the name of the model itself, but also using
 all the parent models following the full inheritance chain. So if all your models
 inherit from ``MyAbstractModel``, you can for example create an override table
 for all your models by creating a ``myproject.tables.MyAbstractModelTable``.
+
+
+Extending autocomplete results
```
Comment (b1rger) [inline docs/source/customization.rst:153]: This duplicates in part the part from line 115-120, so it might make sense to delete those lines

**Review by b1rger — CHANGES_REQUESTED** [review]

File: docs/source/customization.rst (hunk 4 above)
Comment (sennierer) [reply docs/source/customization.rst:158]: ok, removed it

File: docs/source/customization.rst (hunk 23 above)
Comment (sennierer) [reply docs/source/customization.rst:153]: yes, but it contains actual example code....trying to move it in the upper section

Comment (b1rger) [reply docs/source/customization.rst:153]: Yes, I meant removing the older version or merging it together

File: docs/source/customization.rst [hunk 24]
```diff
@@ -17,9 +20,9 @@ The :py:mod:`apis_core.history` module also uses the generic views for
 its models. This means that you can use the generic views for the historical 
 models of your own ontology. E.g. if you have a model ``Person`` in your 
 `apis_ontology.models` module, you can use the generic views for it.
```
Comment (b1rger) [inline docs/source/customization.rst:22]:
```suggestion
``apis_ontology.models`` module, you can use the generic views for it. 
```

File: docs/source/customization.rst [hunk 25]
```diff
@@ -17,9 +20,9 @@ The :py:mod:`apis_core.history` module also uses the generic views for
 its models. This means that you can use the generic views for the historical 
 models of your own ontology. E.g. if you have a model ``Person`` in your 
 `apis_ontology.models` module, you can use the generic views for it. 
-`/apis/apis_ontology.historicalperson/</apis/apis_ontology.historicalperson/>`_ 
+`/apis/apis_ontology.historicalperson/`
```
Comment (b1rger) [inline docs/source/customization.rst:23]:
```suggestion
``/apis/apis_ontology.historicalperson/`` 
```

File: docs/source/customization.rst [hunk 26]
```diff
@@ -17,9 +20,9 @@ The :py:mod:`apis_core.history` module also uses the generic views for
 its models. This means that you can use the generic views for the historical 
 models of your own ontology. E.g. if you have a model ``Person`` in your 
 `apis_ontology.models` module, you can use the generic views for it. 
-`/apis/apis_ontology.historicalperson/</apis/apis_ontology.historicalperson/>`_ 
+`/apis/apis_ontology.historicalperson/` 
 will be the URL for the list view of the historical model. 
-`/apis/api/apis_ontology.historicalperson/</apis/api/apis_ontology.historicalperson/>`_ for the API view.
+`/apis/api/apis_ontology.historicalperson/` for the API view.
```
Comment (b1rger) [inline docs/source/customization.rst:25]:
```suggestion
``/apis/api/apis_ontology.historicalperson/`` for the API view.
```

**Review by b1rger — CHANGES_REQUESTED** [review]

**Review by b1rger — APPROVED** [review]

**Pull request merged**